src/common/framework/gpu/implementation.ts

```diff
@@ -1,19 +1,112 @@
+    module.exports = require('../../../../standalone/third_party/dawn/linux/index.node');
+    return new module.exports.WebGPUWindow({ width, height, title });
+  }
+}
+
+export class GPU {
+  requestAdapter({ window }: { window?: WebGPUWindow }): Promise<GPUAdapter> {
+    module.exports = require('../../../../standalone/third_party/dawn/linux/index.node');
+    return module.exports.GPU.requestAdapter({ window });
+  }
+}
+
+export function getGPU(): Promise<GPU> {
   if (impl) {
     return impl;
   }
 
-  assert(
-    typeof navigator !== 'undefined' && navigator.gpu !== undefined,
-    'No WebGPU implementation found'
-  );
+  let dawn = false;
 
-  impl = navigator.gpu;
+  if (fs.existsSync('standalone/third_party/dawn/linux/index.node')) {
+    dawn = true;
+  }
+
+  if (typeof navigator !== 'undefined' && 'gpu' in navigator) {
+    impl = Promise.resolve(navigator.gpu as GPU);
+  } else if (dawn) {
+    module.exports = require('../../../../standalone/third_party/dawn/linux/index.node');
+    module.exports.GPU.$setPlatform(process.platform);
+    Object.assign(global, module.exports);
+    impl = module.exports.GPU;
+
+    {
+      const { GPUAdapter } = module.exports;
+      GPUAdapter.prototype.requestDevice = function (): Promise<GPUDevice> {
+        console.log('req adapter');
+        return new Promise(resolve => {
+          this._requestDevice().then((device: GPUDevice) => {
+            console.log('req device');
+            resolve(device);
+          });
+        });
+      };
+    }
+
+    // temporary hack to return a promise instead of a callback
+    {
+      const { GPUFence } = module.exports;
+      GPUFence.prototype.onCompletion = function (completionValue: number): Promise<void> {
+        return new Promise(resolve => {
+          setImmediate(() => {
+            this._onCompletion(completionValue, resolve);
+          });
+        });
+      };
+    }
+    {
+      const { GPUBuffer } = module.exports;
+      GPUBuffer.prototype.mapReadAsync = function (): Promise<void> {
+        return new Promise(resolve => {
+          setImmediate(() => {
+            this._mapReadAsync(resolve);
+          });
+        });
+      };
+    }
+    {
+      const { GPUBuffer } = module.exports;
+      GPUBuffer.prototype.mapWriteAsync = function (): Promise<void> {
+        return new Promise(resolve => {
+          setImmediate(() => {
+            this._mapWriteAsync(resolve);
+          });
+        });
+      };
+    }
+    {
+      const { GPUDevice } = module.exports;
+      GPUDevice.prototype.createBufferMappedAsync = function (
+        descriptor: GPUDeviceDescriptor
+      ): Promise<void> {
+        return new Promise(resolve => {
+          setImmediate(() => {
+            this._createBufferMappedAsync(descriptor, resolve);
+          });
+        });
+      };
+    }
+    {
+      const { GPUDevice } = module.exports;
+      GPUDevice.prototype.createBufferMapped = function (
```

Comment:
It is so everything returns a Promise. Please be advised the NAPI addon was written against older versions of WebGPU and will need maintainence moving forward.

Reply:
Ok, I included your changes and made a few suggested changes. Assuming we are on the same page with the results:

- I put some logging in to see where the tests are hanging.
- Seems that CTS is not calling the createBufferMapped polyfill from the updated  implementation.ts file.

Reply:
I'm not aware of any other than the "use strict" thing. However note that there are some JS errors which "use strict" turns from no-ops into exceptions. So it's possible that something done here is no-opping.

Reply:
Where does CTS call device.tick ?

Reply:
It doesn't; there's no tick() in the JS API. Instead the implementation is supposed to figure out when to tick. In Chrome we currently do this dumbly, on a periodic basis, which has more overhead than we'd like. For NAPI I guess there are two options to start with:
- Expose it to JS and call it with setInterval or something
- Set up a setInterval-like timer in C++ during creation of the device

Reply:
Turns out we seem to be exactly where we want to be with this.

The create_mapped.spec test is of course calling the official WebGPU API prototype for createBufferMapped.
However, the implentation file implements a NAPI polyfill which has non-conformant arguments (missing the size).

That is what the polyfill in the implementation file is for right now - to perform this final mappings in preparation for updating the NAPI addon.

Will update the PR as I move forward with the polyfills.

Reply:
Making progress - looks like we are passing the first set of tests!

**> tools/run --verbose webgpu:api,operation,buffers,map_detach:***
[pass] webgpu:api,operation,buffers,map_detach:mapWriteAsync:unmap=true;destroy=false (107.761ms). Log:
[pass] webgpu:api,operation,buffers,map_detach:mapWriteAsync:unmap=true;destroy=true (5.341ms). Log:
[pass] webgpu:api,operation,buffers,map_detach:mapWriteAsync:unmap=false;destroy=true (5.213ms). Log:
[pass] webgpu:api,operation,buffers,map_detach:mapReadAsync:unmap=true;destroy=false (5.396ms). Log:
[pass] webgpu:api,operation,buffers,map_detach:mapReadAsync:unmap=true;destroy=true (5.267ms). Log:
[pass] webgpu:api,operation,buffers,map_detach:mapReadAsync:unmap=false;destroy=true (5.251ms). Log:

Please note the following:

- Node 13+ is minimum requirement for `cts-napi` branch for all 'experimental' NAPI functions to be enabled
- Native error scoping (push and pop) still needs to be implemented in the WebGPU-NAPI addon
- CTS is running the above suite to completion but hangs at the end - still an issue with the event-loop.

Reply:
> * Node 13+ is minimum requirement for `cts-napi` branch for all 'experimental' NAPI functions to be enabled

ah, this is certainly why I couldn't run the plugin before.

Reply:
... well, maybe not. Same error with Node 14.6.0.

Reply:
> However, the implentation file implements a NAPI polyfill which has non-conformant arguments (missing the size). CTS should really report an error instead of just blocking should it not?

Yes, the test should fail (is that not what happens?)

Reply:
I am not seeing any errors - that is my issue right now. CTS just hangs when create_mapped.spec.ts tries to call the createBufferMapped method of GPUDevice. Other tests pass (above), but those don't require prototyping (called directly from the WebGPU-NAPI addon).

Thinking this is an dev enviroment or version mismatch issue maybe - tsconfig.json settings?
Are you still seeing an error and if so could you post it here?

Reply:
Seems like some kind of issue in the binary (maybe issue with compatibility with my node?)
It's the same error as before: https://github.com/gpuweb/cts/pull/232#issuecomment-657890026

Reply:
I am running with same node version 13.14.0 and 14.6.0 with no such error, passing mapWriteAsync.
Which version of Ubuntu are you using?

Reply:
Didn't work. Unfortunately I'm really swamped with work so I don't have time to keep investigating this for now.

BTW my system isn't the supported ubuntu, though from the error I don't suspect that's the problem (I'm on some version of debian testing). First thing on the list if I get back to investigating this though.

Reply:
I am backing out your code changes for now from our cts-napi banch as you advised, and proceeding with our original integration code. We really have to be on the same page in terms of the dev environment to be able to collaborate (Ubuntu 18 or 20 LTS is current target).

Reply:
Sorry that patch was broken; like I said it was just a sketch; I hadn't intend for you to merge it as-is since I hadn't been able to test it.

Reply:
Looking into this with Felix. His RT WebGPU-NAPI is curently built against a fork of Dawn that needs to be rebased with lastest upstream changes of Dawn.

Our WebGPU-NAPI is based directly on the RT version (with RT disabled) and currently can build against the July 10 image of Dawn. I am working to get it rebased to the  current upstream Dawn.

I am now getting the same error you met with, so we are on the same page again. This error does not involve upstream changes - WebGPU-NAPI does not have comprehensive test suites, which is why we are doing this CTS integration.

Reply:
Rewrote createBufferMapped in [WebGPU_NAPI](https://github.com/Kings-Distributed-Systems/webgpu-napi/commit/c0f8b3b9543034444c48b20af24510c21ebb3c56#diff-0152a20671fbb9f882b5ab3670742826).

Seems to be workingin the addon but there is a problem in the typescript call to it in [map_detach.spec.ts](https://github.com/Kings-Distributed-Systems/cts/blob/cts-napi/src/webgpu/napi/operation/buffers/map_detach.spec.ts)


```
tools/run --verbose --debug webgpu:napi,operation,buffers,create_mapped:*
[fail] webgpu:napi,operation,buffers,create_mapped:createBufferMapped:size=12;mappable=false (133.607ms). Log:
  - EXCEPTION: Cannot convert undefined or null to object
    TypeError: Cannot convert undefined or null to object
        at RunCaseSpecific.fn (/home/dcerisano/git/cts/src/webgpu/napi/operation/buffers/create_mapped.spec.ts:26:44)
[fail] webgpu:napi,operation,buffers,create_mapped:createBufferMapped:size=12;mappable=true (1.376ms). Log:
  - EXCEPTION: Cannot convert undefined or null to object
    TypeError: Cannot convert undefined or null to object
        at RunCaseSpecific.fn (/home/dcerisano/git/cts/src/webgpu/napi/operation/buffers/create_mapped.spec.ts:26:44)
[fail] webgpu:napi,operation,buffers,create_mapped:createBufferMapped:size=524288;mappable=false (0.242ms). Log:
  - EXCEPTION: Cannot convert undefined or null to object
    TypeError: Cannot convert undefined or null to object
        at RunCaseSpecific.fn (/home/dcerisano/git/cts/src/webgpu/napi/operation/buffers/create_mapped.spec.ts:26:44)
[fail] webgpu:napi,operation,buffers,create_mapped:createBufferMapped:size=524288;mappable=true (0.245ms). Log:
  - EXCEPTION: Cannot convert undefined or null to object
    TypeError: Cannot convert undefined or null to object
        at RunCaseSpecific.fn (/home/dcerisano/git/cts/src/webgpu/napi/operation/buffers/create_mapped.spec.ts:26:44)

** Failures **
[fail] webgpu:napi,operation,buffers,create_mapped:createBufferMapped:size=12;mappable=false (133.607ms). Log:
  - EXCEPTION: Cannot convert undefined or null to object
    TypeError: Cannot convert undefined or null to object
        at RunCaseSpecific.fn (/home/dcerisano/git/cts/src/webgpu/napi/operation/buffers/create_mapped.spec.ts:26:44)
[fail] webgpu:napi,operation,buffers,create_mapped:createBufferMapped:size=12;mappable=true (1.376ms). Log:
  - EXCEPTION: Cannot convert undefined or null to object
    TypeError: Cannot convert undefined or null to object
        at RunCaseSpecific.fn (/home/dcerisano/git/cts/src/webgpu/napi/operation/buffers/create_mapped.spec.ts:26:44)
[fail] webgpu:napi,operation,buffers,create_mapped:createBufferMapped:size=524288;mappable=false (0.242ms). Log:
  - EXCEPTION: Cannot convert undefined or null to object
    TypeError: Cannot convert undefined or null to object
        at RunCaseSpecific.fn (/home/dcerisano/git/cts/src/webgpu/napi/operation/buffers/create_mapped.spec.ts:26:44)
[fail] webgpu:napi,operation,buffers,create_mapped:createBufferMapped:size=524288;mappable=true (0.245ms). Log:
  - EXCEPTION: Cannot convert undefined or null to object
    TypeError: Cannot convert undefined or null to object
        at RunCaseSpecific.fn (/home/dcerisano/git/cts/src/webgpu/napi/operation/buffers/create_mapped.spec.ts:26:44)

** Summary **
Passed  w/o warnings = 0 / 4 =   0.00%
Passed with warnings = 0 / 4 =   0.00%
Skipped              = 0 / 4 =   0.00%
Failed               = 4 / 4 = 100.00%
```